File: modules/expo-bluesky-oauth-client/ios/CryptoUtil.swift
import CryptoKit
import JOSESwift
import ExpoModulesCore

class CryptoUtil {
  // The equivalent of crypto.subtle.digest() with JS on web
  public static func digest(data: Data) -> Data {
    let hash = SHA256.hash(data: data)
    return Data(hash)
  }
  
  public static func getRandomValues(byteLength: Int) -> Data {
    let bytes = (0..<byteLength).map { _ in UInt8.random(in: UInt8.min...UInt8.max) }
    return Data(bytes)
  }
  
  public static func generateKeyPair() throws -> [String:String] {
    let keyIdString = UUID().uuidString
    
    let privateKey = P256.Signing.PrivateKey()
    let publicKey = privateKey.publicKey
    
    let x = publicKey.x963Representation[1..<33].base64URLEncodedString()
    let y = publicKey.x963Representation[33...].base64URLEncodedString()
    let d = privateKey.rawRepresentation.base64URLEncodedString()
        
    let publicJWK = JWK(kty: "EC", use: "sig", crv: "P-256", kid: keyIdString, x: x, y: y, alg: "ES256")
    let privateJWK = JWK(kty: "EC", use: "sig", crv: "P-256", kid: keyIdString, x: x, y: y, d: d, alg: "ES256")
    
    return [
      "privateKey": privateJWK.toJson(),
      "publicKey": publicJWK.toJson()
    ]
  }
}

extension Data {
  func base64URLEncodedString() -> String {
    return self.base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
  }
}

extension String {
  func toField() -> Field<String> {
    return Field(wrappedValue: self)
  }
  func toNullableField() -> Field<String?> {
    return Field(wrappedValue: self)
  }
}

File: modules/expo-bluesky-oauth-client/android/src/main/java/expo/modules/blueskyoauthclient/CryptoUtil.kt
package expo.modules.blueskyoauthclient

import com.nimbusds.jose.Algorithm
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.interfaces.ECPublicKey
import java.security.interfaces.ECPrivateKey
import com.nimbusds.jose.jwk.Curve
import com.nimbusds.jose.jwk.ECKey
import com.nimbusds.jose.jwk.KeyUse
import java.util.UUID

class CryptoUtil {
  fun digest(data: ByteArray, algorithmName: String): ByteArray {
    if(algorithmName != "sha256") throw Exception("Unsupported algorithm")
    val digest = MessageDigest.getInstance("sha256")
    return digest.digest(data)
  }

  fun getRandomValues(byteLength: Int): ByteArray {
    val random = ByteArray(byteLength)
    java.security.SecureRandom().nextBytes(random)
    return random
  }

  fun generateKeyPair(): Any {
    val keyIdString = UUID.randomUUID().toString()

    val keyPairGen = KeyPairGenerator.getInstance("EC")
    keyPairGen.initialize(Curve.P_256.toECParameterSpec())
    val keyPair = keyPairGen.generateKeyPair()

    val publicKey = keyPair.public as ECPublicKey
    val privateKey = keyPair.private as ECPrivateKey

    val publicJwk = ECKey.Builder(Curve.P_256, publicKey)
      .keyUse(KeyUse.SIGNATURE)
      .algorithm(Algorithm.parse("ES256"))
      .keyID(keyIdString)
      .build()
    val privateJwk = ECKey.Builder(Curve.P_256, publicKey)
      .privateKey(privateKey)
      .keyUse(KeyUse.SIGNATURE)
      .keyID(keyIdString)
      .algorithm(Algorithm.parse("ES256"))
      .build()


    return JWKPair(
      JWK(
        alg = privateJwk.algorithm.toString(),
        kty = privateJwk.keyType.toString(),
        crv = privateJwk.curve.toString(),
        x = privateJwk.x.toString(),
        y = privateJwk.y.toString(),
        d = privateJwk.d.toString(),
        use = privateJwk.keyUse.toString(),
        kid = privateJwk.keyID
      ),
      JWK(
        alg = publicJwk.algorithm.toString(),
        kty = publicJwk.keyType.toString(),
        crv = publicJwk.curve.toString(),
        x = publicJwk.x.toString(),
        y = publicJwk.y.toString(),
        use = publicJwk.keyUse.toString(),
        kid = publicJwk.keyID
      )
    )
  }
}
